src/main/java/org/opensearch/knn/index/mapper/EngineFieldMapper.java

```diff
@@ -0,0 +1,286 @@
+        KNNMethodConfigContext knnMethodConfigContext,
+        MultiFields multiFields,
+        CopyTo copyTo,
+        Explicit<Boolean> ignoreMalformed,
+        boolean stored,
+        boolean hasDocValues,
+        OriginalMappingParameters originalMappingParameters
+    ) {
+        KNNMethodContext methodContext = originalMappingParameters.getResolvedKnnMethodContext();
+        KNNLibraryIndexingContext libraryContext = methodContext.getKnnEngine()
+            .getKNNLibraryIndexingContext(methodContext, knnMethodConfigContext);
+        boolean isLuceneEngine = KNNEngine.LUCENE.equals(methodContext.getKnnEngine());
+
+        KNNVectorFieldType mappedFieldType = new KNNVectorFieldType(
+            fullname,
+            metaValue,
+            knnMethodConfigContext.getVectorDataType(),
+            new KNNMappingConfig() {
+                @Override
+                public Optional<KNNMethodContext> getKnnMethodContext() {
+                    return Optional.of(methodContext);
+                }
+
+                @Override
+                public int getDimension() {
+                    return knnMethodConfigContext.getDimension();
+                }
+
+                @Override
+                public Mode getMode() {
+                    return Mode.fromName(originalMappingParameters.getMode());
+                }
+
+                @Override
+                public CompressionLevel getCompressionLevel() {
+                    return knnMethodConfigContext.getCompressionLevel();
+                }
+
+                @Override
+                public Version getIndexCreatedVersion() {
+                    return knnMethodConfigContext.getVersionCreated();
+                }
+
+                @Override
+                public QuantizationConfig getQuantizationConfig() {
+                    return Optional.ofNullable(libraryContext)
+                        .map(KNNLibraryIndexingContext::getQuantizationConfig)
+                        .orElse(QuantizationConfig.EMPTY);
+                }
+
+                @Override
+                public KNNLibraryIndexingContext getKnnLibraryIndexingContext() {
+                    return libraryContext;
+                }
+            }
+        );
+
+        return new EngineFieldMapper(
+            simpleName,
+            mappedFieldType,
+            multiFields,
+            copyTo,
+            ignoreMalformed,
+            stored,
+            hasDocValues,
+            knnMethodConfigContext,
+            originalMappingParameters,
+            isLuceneEngine
+        );
+    }
+
+    private EngineFieldMapper(
+        String name,
+        KNNVectorFieldType mappedFieldType,
+        MultiFields multiFields,
+        CopyTo copyTo,
+        Explicit<Boolean> ignoreMalformed,
+        boolean stored,
+        boolean hasDocValues,
+        KNNMethodConfigContext knnMethodConfigContext,
+        OriginalMappingParameters originalMappingParameters,
+        boolean isLuceneEngine
```

Comment:
this defeats the purpose of having a single FieldMapper. If we are not getting rid of this boolean flag, then I would suggest lets not have a single EngineFieldMapper class. Because this boolean flag will create bifercations in the code, and then tomorrow if we bring in a new engine then will we will be creating a new boolean?

Reply:
I have that as a followup task. The initial commit here just lays out the structure and default values.